contracts/test/MarionetteOld.sol

```diff
@@ -0,0 +1,188 @@
+    using AddressUpgradeable for address payable;
+
+    struct FunctionCall {
+        address receiver;
+        uint value;
+        bytes data;
+    }
+
+    bytes32 public constant IMA_ROLE = keccak256("IMA_ROLE");
+    bytes32 public constant PUPPETEER_ROLE = keccak256("PUPPETEER_ROLE");
+    string public constant ACCESS_VIOLATION = "Access violation";
+
+    event EtherReceived(
+        address sender,
+        uint amount
+    );
+
+    event EtherSent(
+        address receiver,
+        uint amount
+    );
+
+    event FunctionCallResult (
+        bytes output
+    );
+
+    receive() external payable override {
+        emit EtherReceived(msg.sender, msg.value);
+    }
+
+    function initialize(address owner, address ima) external override initializer {
+        AccessControlEnumerableUpgradeable.__AccessControlEnumerable_init();
+        AccessControlUpgradeable._setupRole(DEFAULT_ADMIN_ROLE, address(this));
+        AccessControlUpgradeable._setupRole(PUPPETEER_ROLE, owner);
+        if (ima != address(0)) {
+            AccessControlUpgradeable._setupRole(IMA_ROLE, ima);
+        }
+    }
+
+    function postMessage(
+        bytes32,
+        address sender,
+        bytes calldata data
+    )
+        external
+        override
+        returns (address)
+    {
+        require(hasRole(IMA_ROLE, msg.sender), "Sender is not IMA");
+        require(hasRole(PUPPETEER_ROLE, sender), ACCESS_VIOLATION);
+
+        FunctionCall memory functionCall = _parseFunctionCall(data);
+
+        bytes memory output = _doCall(payable(functionCall.receiver), functionCall.value, functionCall.data);        
+        emit FunctionCallResult(output);
+
+        return address(0);
+    }
+
+    function execute(
+        address payable target,
+        uint value,
+        bytes calldata data
+    )
+        external
+        payable
+        override
+        returns (bytes memory)
+    {
+        require(hasRole(PUPPETEER_ROLE, msg.sender), ACCESS_VIOLATION);
+
+        return _doCall(target, value, data);
+    }
+
+    function sendEth(address payable target, uint value) external payable override {
+        require(hasRole(PUPPETEER_ROLE, msg.sender), ACCESS_VIOLATION);
+
+        _doCall(target, value, "0x");
+    }
+
+    function encodeFunctionCall(
+        address receiver,
+        uint value,
+        bytes calldata data
+    )
+        external
+        pure
+        override
+        returns (bytes memory)
+    {
+        return abi.encode(receiver, value, data);
+    }
+
+    // private
+
+       function _doCall(address payable target, uint value, bytes memory data) private returns (bytes memory) {
+        if (msg.value > 0) {
+            emit EtherReceived(msg.sender, msg.value);
+        }
+
+        if (value > 0) {
+            emit EtherSent(target, value);
+        }
+
+        if (target.isContract()) {
+            if (data.length >= 4) {
+                if (_getHashedSelector(data) == _multiSendHashed()) {
```

Comment:
I saw this changes in new contract, I thought this contract would be a previous contract